Add support for the v1 alerts api.

hhellyer opened the pull request:
This PR adds support for the alerts api.

- Adds code to call the new alerts api.
- Adds testcases.
- Adds example.

hhellyer committed:
Add support for the v1beta1 alerts api.
- Add code to call api.
- Add testcases.
- Add example.

apantelopoulosNS1 commented on rest/model/alerting/alert.go:
```diff
@@ -0,0 +1,46 @@
+package alerting
```
oh nice, more models! its not like that we did not have enough already :D

apantelopoulosNS1 replied:
not a jab at the PR but rather than the fact that we have a 1000 models!

apantelopoulosNS1 commented on rest/_examples/alerts.go:
```diff
@@ -0,0 +1,132 @@
+var client *api.Client
+
+// Helper that initializes rest api client from environment variable.
+func init() {
```
Is this called in the example somewhere?

hhellyer replied:
If I'm 100% honest this is here because I started by copying one of the other examples and wanted it to follow our existing pattern.
It will always be called as go ensures the `init()` function for a file is always run:
https://go.dev/doc/effective_go#init
(The internet seems rather divided on whether it's a bad thing to use which in general but it's probably ok for samples like this.)

hhellyer committed:
Update for alert center GA.
- Use alerting/v1 in REST calls.
- Update changelog and client version.
- Minor fix to example to create secondary zone correctly.

hhellyer committed:
Fix pagination testcase.

shane-ns1 commented on rest/_examples/zones.go:
```diff
@@ -62,7 +62,7 @@ func main() {
 
 	// Add an A record with a single static answer.
-	orchidRec := dns.NewRecord(domain, "orchid", "A")
+	orchidRec := dns.NewRecord(domain, "orchid", "A", nil, nil)
```
Is `dns.NewRecord()` a call that is part of the Go API? If so, extending it to require additional parameters is a breaking change, meaning we'd have to go from version 2.12.2 to version 3.

If we don't want version 3 (which probably we do not?) and this is part of our Go API, then we should consider adding a new function instead. `dns.NewRecordEx()` or something like that, if we follow Windows-style naming conventions, or `dns.NewRecord2()` if we follow POSIX-style naming convention (I prefer the Windows apporach in this case).

hhellyer replied:
Ah - I should have put a note in the PR.
I only changed this because dns.NewRecord has taken 5 params since https://github.com/ns1/ns1-go/pull/220 some time ago and the example was already broken.
I probably should have raised a separate issue. I can move this to a different issue if you like.

shane-ns1 replied:
No worries. We should have made a major version bump or added a new "constructor" at that time, but we are where we are. :laughing:

Since this is just an example fix from your PR, I think we're good.

shane-ns1 commented on rest/model/alerting/alert.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+// TODO - Allow zones/notifier lists to be passed as structs?
+//   - Multiple constructors?
```
I'm not a fan of TODO comments in our code. If there is a TODO file for our Go API, then I guess we can put it there. Otherwise we can make a GitHub issue with this idea.

apantelopoulosNS1 commented on rest/client.go:
```diff
@@ -13,9 +13,9 @@ import (
+	clientVersion = "2.13.0"
 
-	defaultBase                   = "http://localhost:80"
+	defaultBase                   = "https://api.nsone.net"
```
why did you change this?

shane-ns1 commented on rest/alert.go:
```diff
@@ -0,0 +1,207 @@
+	return resp, nil
+}
+
+// Delete immediately deletes an existing alert.
```
The word "immediately" might make my Spidey-Sense tingle if I was a customer reading this. Why do we say "immediately"? Is there a _non_-immediate way to delete an alert? Do other commands not work immediately? :laughing:

hhellyer replied:
I copied that from the notifier list sdk code - I was trying to blend in. Removed.

shane-ns1 commented on rest/alert_test.go:
```diff
@@ -0,0 +1,562 @@
+	api "gopkg.in/ns1/ns1-go.v2/rest"
+)
+
+func strPtr(str string) *string {
```
Annoying that such a function is needed. I should add this to my "Why I Hate Go" presentation!

shane-ns1 commented on rest/alert_test.go:
```diff
@@ -0,0 +1,562 @@
+			compareAlerts(t, alertResponse, &alertToCreate)
+		})
+
+		// TODO Add test for error response when updating fields that can't be changed.
```
Is this difficult? Why isn't it in the PR?

hhellyer replied:
Removed, it's not difficult it's also not an issue on alert creation - I copied this from either update or replace which does test this.

shane-ns1 reviewed (changes requested):
I have a few questions, nothing very critical.

I am concerned about possibly changing the API without bumping the major version number, but I maybe this concern is misplaced.

hhellyer committed:
Fix review comments.

shane-ns1 reviewed (approved):
Looks good. Thanks for the attention to detail. :+1:
